File: refactor.js
const fs = require('fs');
const path = require('path');

const directory = 'c:/Users/justi/OneDrive/Documents/melora-tunes/components';
const appDirectory = 'c:/Users/justi/OneDrive/Documents/melora-tunes/app';
const hookImport = 'import { useAudioProgress } from "@/hooks/use-audio-progress";\n';

function processDir(dirPath) {
    if (!fs.existsSync(dirPath)) return;
    const files = fs.readdirSync(dirPath, { withFileTypes: true });
    for (const file of files) {
        const fullPath = path.join(dirPath, file.name);
        if (file.isDirectory()) {
            processDir(fullPath);
        } else if (file.name.endsWith('.tsx')) {
            let content = fs.readFileSync(fullPath, 'utf8');

            if (content.includes('usePlayback') && /\{[^\}]*\bprogress\b[^\}]*\}\s*=\s*usePlayback\(/.test(content)) {
                if (!content.includes('useAudioProgress')) {
                    const importPattern = /^import .*$/gm;
                    let lastMatch = null;
                    let match;
                    while ((match = importPattern.exec(content)) !== null) {
                        lastMatch = match;
                    }
                    if (lastMatch) {
                        const insertPos = lastMatch.index + lastMatch[0].length + 1;
                        content = content.slice(0, insertPos) + hookImport + content.slice(insertPos);
                    } else {
                        content = hookImport + content;
                    }
                }

                content = content.replace(/\{([^\}]*\bprogress\b[^\}]*)\}\s*=\s*usePlayback\(\)/g, (match, inner) => {
                    let innerNew = inner.replace(/,\s*progress\b|\bprogress\s*,?/g, '').trim();
                    if (!innerNew) {
                        return `usePlayback();\n    const { progress } = useAudioProgress()`;
                    }
                    return `{ ${innerNew} } = usePlayback();\n    const { progress } = useAudioProgress()`;
                });

                fs.writeFileSync(fullPath, content, 'utf8');
                console.log('Updated ' + file.name);
            }
        }
    }
}
processDir(directory);
processDir(appDirectory);

File: refactor.py
import os
import re

directory = 'c:/Users/justi/OneDrive/Documents/melora-tunes/components'
app_directory = 'c:/Users/justi/OneDrive/Documents/melora-tunes/app'
hook_import = 'import { useAudioProgress } from "@/hooks/use-audio-progress";\n'

def process_dir(dir_path):
    for root, _, files in os.walk(dir_path):
        for file in files:
            if file.endswith('.tsx'):
                path = os.path.join(root, file)
                with open(path, 'r', encoding='utf-8') as f:
                    content = f.read()
                
                # If uses usePlayback and extracts progress
                if 'usePlayback' in content and re.search(r'\{[^\}]*\bprogress\b[^\}]*\}\s*=\s*usePlayback\(\)', content):
                    # Add import if missing
                    if 'useAudioProgress' not in content:
                        imports = list(re.finditer(r'^import .*$', content, re.MULTILINE))
                        if imports:
                            last_import = imports[-1]
                            insert_pos = last_import.end() + 1
                            content = content[:insert_pos] + hook_import + content[insert_pos:]
                        else:
                            content = hook_import + content
                    
                    # Replace the destructuring to remove progress
                    def repl(m):
                        inner = m.group(1)
                        # remove 'progress' and possible commas
                        inner_new = re.sub(r',\s*progress\b|\bprogress\s*,?', '', inner).strip()
                        if not inner_new:
                            return f'usePlayback();\n    const {{ progress }} = useAudioProgress();'
                        return f'{{ {inner_new} }} = usePlayback();\n    const {{ progress }} = useAudioProgress();'
                        
                    content = re.sub(r'\{([^\}]*\bprogress\b[^\}]*)\}\s*=\s*usePlayback\(\);?', repl, content)
                    
                    with open(path, 'w', encoding='utf-8') as f:
                        f.write(content)
                    print(f'Updated {file}')

process_dir(directory)
process_dir(app_directory)
